### PR TITLE
Unstuck Snowbridge

### PR DESCRIPTION
**Pull request by claravanstaden**
Upgrades Snowbridge with an Ethereum client fixes: 
- https://github.com/paritytech/polkadot-sdk/pull/4504
- https://github.com/paritytech/polkadot-sdk/pull/4482

Adds a migration to reset the Ethereum checkpoint. Will be reset again to the moment recent checkpoint before merged.

TODO:
- [x] Check migration works
- [x] Check converting Sync Committee to prepared keys is OK for a migration

### CODE REVIEW

File: system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_unstuck.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+
+pub struct UnstuckSnowbridge;
+
+impl OnRuntimeUpgrade for UnstuckSnowbridge {
```
Comment (acatangiu) [inline system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_unstuck.rs:44]: You also need to add it to the list of migrations ([example](https://github.com/polkadot-fellows/runtimes/blob/c2886b940944ba96c87b6e07954e0d945b9a67bd/system-parachains/bridge-hubs/bridge-hub-polkadot/src/lib.rs#L151))

Comment (claravanstaden) [reply system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_unstuck.rs:44]: Thanks for the reminder! https://github.com/polkadot-fellows/runtimes/pull/313/commits/b13f7a97601e91759ebd8c5aabfb3a334f03b418

File: system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_unstuck.rs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        );
+        LatestFinalizedBlockRoot::<Runtime>::set(NEW_CHECKPOINT_HEADER_ROOT.into());
+
+        RocksDbWeight::get().reads_writes(6, 6)
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() -> Result<Vec<u8>, TryRuntimeError> {
+        if !is_bridge_stuck() {
+            log::info!(target: LOG_TARGET, "Going to unstuck bridge.");
```
Comment (acatangiu) [inline system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_unstuck.rs:69]:
```suggestion
            log::info!(target: LOG_TARGET, "Going to unstuck Snowbridge.");
```

File: system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_unstuck.rs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        } else {
+            log::info!(target: LOG_TARGET, "Bridge is not stuck.");
```
Comment (acatangiu) [inline system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_unstuck.rs:71]:
```suggestion
            log::info!(target: LOG_TARGET, "Snowbridge is not stuck.");
```

File: system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_unstuck.rs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        }
+        Ok(vec![])
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn post_upgrade(_: Vec<u8>) -> Result<(), TryRuntimeError> {
+        frame_support::ensure!(!is_bridge_stuck(), "Bridge is still stuck.");
```
Comment (acatangiu) [inline system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_unstuck.rs:78]:
```suggestion
        frame_support::ensure!(!is_bridge_stuck(), "Snowbridge is still stuck.");
```

File: system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_unstuck.rs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        Ok(())
+    }
+}
+
+fn is_bridge_stuck() -> bool {
+    LatestFinalizedBlockRoot::<Runtime>::get() == LAST_IMPORTED_BEACON_HEADER.into()
```
Comment (acatangiu) [inline system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_unstuck.rs:84]: We expect that this `LAST_IMPORTED_BEACON_HEADER` will not be reimported, yes?

(I guess the checkpoint is from a block newer than that)

**Comment by acatangiu** [comment]
> @acatangiu the migration is roughly doing the same as the `force_checkpoint` extrinsic. It looks like an expensive operation compared to the P<>K bridge migration - the weight is [Weight::from_parts(101_492_831_000, 0)](https://github.com/polkadot-fellows/runtimes/blob/main/system-parachains/bridge-hubs/bridge-hub-polkadot/src/weights/snowbridge_pallet_ethereum_client.rs#L72). Do you think this will be a problem? We'll test of course but just wanted to also ask your opinion.

Should be ok as long as it fits a block.

**Comment by vgeddes** [comment]
> > @acatangiu the migration is roughly doing the same as the `force_checkpoint` extrinsic. It looks like an expensive operation compared to the P<>K bridge migration - the weight is [Weight::from_parts(101_492_831_000, 0)](https://github.com/polkadot-fellows/runtimes/blob/main/system-parachains/bridge-hubs/bridge-hub-polkadot/src/weights/snowbridge_pallet_ethereum_client.rs#L72). Do you think this will be a problem? We'll test of course but just wanted to also ask your opinion.
> 
> Should be ok as long as it fits a block.

Yeah this is fine, is well below the block limit. And we know it works because our initial governance proposal already executed a `force_checkpoint` successfully on BridgeHub.

File: system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_unstuck.rs [hunk 6]
```diff
@@ -53,23 +48,26 @@ impl OnRuntimeUpgrade for UnstuckSnowbridge {
 			return RocksDbWeight::get().reads(1);
 		}
 
-		log::info!(target: LOG_TARGET, "Updating beacon checkpoint set to unstuck bridge.");
+		log::info!(target: LOG_TARGET, "Updating beacon checkpoint to unstuck beacon clieny");
```
Comment (claravanstaden) [inline system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_unstuck.rs:51]: *beacon client 😄

**Review by vgeddes — APPROVED** [review]
+1

**Review by acatangiu — APPROVED** [review]
Migration code looks good, also validated that the new checkpoint leads to https://beaconscan.com/slot/9094528.

**Review by alistair-singh — APPROVED** [review]

**Review by joepetrowski — APPROVED** [review]

**Pull request merged**